From f6e419bea5314d2d507b6b45c6871e018806671c Mon Sep 17 00:00:00 2001
From: Nicolin Chen <nicolinc@nvidia.com>
Date: Thu, 7 Jul 2022 00:15:58 -0700
Subject: [PATCH 136/141] iommu/arm-smmu-v3: Implement arm_smmu_get_msi_domain

This allows dma-iommu core to get the correct msi_cookie for a nested
setup, where the msi_cookie is in the S2 domain while the device gets
attached to the S1 domain.

Change-Id: Ib74266245a6c1459fb08d4aebc940f0c759f2410
Signed-off-by: Nicolin Chen <nicolinc@nvidia.com>

12	0	drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c

diff --git a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
index 6decae94b5bb..360ad2d3c50b 100644
--- a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
+++ b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
@@ -2917,6 +2917,17 @@ struct iommu_domain *arm_smmu_nested_domain_alloc(struct iommu_domain *s2_domain
 	return domain;
 }
 
+static struct iommu_domain *arm_smmu_get_msi_domain(struct device *dev)
+{
+	struct arm_smmu_master *master = dev_iommu_priv_get(dev);
+	struct arm_smmu_domain *smmu_domain = master->domain;
+
+	if (smmu_domain->s2)
+		return &smmu_domain->s2->domain;
+
+	return &smmu_domain->domain;
+}
+
 static bool arm_smmu_dev_has_feature(struct device *dev,
 				     enum iommu_dev_features feat)
 {
@@ -3000,6 +3011,7 @@ static struct iommu_ops arm_smmu_ops = {
 	.capable		= arm_smmu_capable,
 	.domain_alloc		= arm_smmu_domain_alloc,
 	.nested_domain_alloc	= arm_smmu_nested_domain_alloc,
+	.get_msi_domain		= arm_smmu_get_msi_domain,
 	.probe_device		= arm_smmu_probe_device,
 	.release_device		= arm_smmu_release_device,
 	.device_group		= arm_smmu_device_group,
-- 
2.34.1

